File: scripts/lib/session_lifecycle.py
"""会话生命周期 — 与 session-lifecycle.mjs 对应。"""
from __future__ import annotations

from typing import Any, Callable


def resolve_app_id(cfg: dict[str, Any], app_id: str | None = None) -> str | None:
    return app_id or cfg.get("appId") or cfg.get("app_id")


def should_keep_target(opts: dict[str, Any] | None = None) -> bool:
    o = opts or {}
    return bool(o.get("keepApp") or o.get("keepBrowser") or o.get("keepTarget"))


def close_with_target(
    *,
    platform: str,
    cfg: dict[str, Any],
    run: Callable[[str, dict | None], None],
    session_close: Callable[[], None],
    opts: dict[str, Any] | None = None,
    app_id: str | None = None,
) -> None:
    """close 默认先 exit 再关会话；keepApp/keepBrowser/keepTarget 时仅关会话。"""
    if not should_keep_target(opts):
        exit_target(platform=platform, cfg=cfg, run=run, session_close=session_close, app_id=app_id)
        if platform == "web":
            return
    session_close()


def exit_target(
    *,
    platform: str,
    cfg: dict[str, Any],
    run: Callable[[str, dict | None], None],
    session_close: Callable[[], None],
    app_id: str | None = None,
) -> None:
    if platform == "web":
        session_close()
        return
    aid = resolve_app_id(cfg, app_id)
    if not aid:
        return
    run("exitApp", {"appId": aid})

File: scripts/lib/session-lifecycle.mjs
/**
 * 会话生命周期：exit（关 App/浏览器）与 close（会话 + 默认关目标）
 */

/** @param {object} cfg @param {string} [appId] */
export function resolveAppId(cfg, appId) {
  return appId ?? cfg.appId ?? cfg.app_id;
}

/** @param {object} [opts] */
export function shouldKeepTarget(opts = {}) {
  return opts.keepApp === true || opts.keepBrowser === true || opts.keepTarget === true;
}

/**
 * 结束用户可见目标：移动 force-stop App；Web 关闭浏览器会话
 * @param {"web"|"android"|"harmony"|"ios"} platform
 * @param {object} cfg
 * @param {{ run?: (cmd: string, extra?: object) => Promise<unknown>, sessionClose: () => Promise<unknown> }} deps
 */
export function createTargetExit(platform, cfg, { run, sessionClose }) {
  if (platform === "web") {
    return async () => sessionClose();
  }
  return async (appId) => {
    const id = resolveAppId(cfg, appId);
    if (!id) return;
    await run("exitApp", { appId: id });
  };
}

/**
 * @param {"web"|"android"|"harmony"|"ios"} platform
 * @param {() => Promise<unknown>} targetExit
 * @param {() => Promise<unknown>} sessionClose
 */
export function createSessionClose(platform, targetExit, sessionClose) {
  const exit = async (appId) => targetExit(appId);

  const close = async (opts = {}) => {
    if (!shouldKeepTarget(opts)) {
      await exit();
      if (platform === "web") return;
    }
    await sessionClose();
  };

  return { exit, close };
}
